File: ios/Plugin/Transport/TcpTransport.swift
import Foundation
import Network

/// Transport TCP (RAW / port 9100) basé sur Network.framework (NWConnection).
///
/// ⚠️ iOS : la première connexion à une IP locale déclenche la pop-up système
/// "Réseau local". Il FAUT déclarer NSLocalNetworkUsageDescription dans Info.plist.
final class TcpTransport {

    private let host: String
    private let port: UInt16
    private var connection: NWConnection?
    private let queue = DispatchQueue(label: "thermalprinter.tcp")

    private(set) var isOpen = false

    init(host: String, port: UInt16 = 9100) {
        self.host = host
        self.port = port
    }

    func open(timeoutMs: Int) async throws {
        if isOpen { return }
        let conn = NWConnection(
            host: NWEndpoint.Host(host),
            port: NWEndpoint.Port(rawValue: port)!,
            using: .tcp
        )
        self.connection = conn

        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            var resumed = false
            func finish(_ result: Result<Void, Error>) {
                if resumed { return }
                resumed = true
                cont.resume(with: result)
            }
            conn.stateUpdateHandler = { [weak self] state in
                switch state {
                case .ready:
                    self?.isOpen = true
                    finish(.success(()))
                case .failed(let err):
                    finish(.failure(PrinterError(.CONNECTION_FAILED, "TCP \(self?.host ?? "")", detail: err.localizedDescription, retryable: true)))
                case .cancelled:
                    finish(.failure(PrinterError(.CONNECTION_FAILED, "TCP annulé")))
                default: break
                }
            }
            conn.start(queue: queue)
            queue.asyncAfter(deadline: .now() + .milliseconds(timeoutMs)) {
                if !resumed {
                    conn.cancel()
                    finish(.failure(PrinterError(.TIMEOUT, "Timeout TCP \(self.host):\(self.port)", retryable: true)))
                }
            }
        }
    }

    func write(_ bytes: [UInt8]) async throws {
        guard let conn = connection, isOpen else {
            throw PrinterError(.CONNECTION_FAILED, "Socket TCP non ouvert")
        }
        let data = Data(bytes)
        // Envoi par chunks pour ménager les petits buffers imprimante.
        let chunkSize = 4096
        var offset = 0
        while offset < data.count {
            let end = min(offset + chunkSize, data.count)
            let chunk = data.subdata(in: offset..<end)
            try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
                conn.send(content: chunk, completion: .contentProcessed { error in
                    if let error = error {
                        cont.resume(throwing: PrinterError(.PRINT_FAILED, "Écriture TCP échouée", detail: error.localizedDescription, retryable: true))
                    } else {
                        cont.resume()
                    }
                })
            }
            offset = end
        }
    }

    func close() {
        connection?.cancel()
        connection = nil
        isOpen = false
    }
}

File: android/src/main/java/com/resto/thermalprinter/transport/TcpTransport.kt
package com.resto.thermalprinter.transport

import com.resto.thermalprinter.model.ErrorCode
import com.resto.thermalprinter.model.PrinterException
import java.io.InputStream
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.Socket
import java.net.SocketTimeoutException

/**
 * Transport TCP brut (RAW / port 9100). Utilisé par EscPosAdapter et RawTcpAdapter
 * en Wi-Fi/Ethernet.
 *
 * @param host adresse IP/hostname
 * @param port port (défaut 9100)
 */
class TcpTransport(
    private val host: String,
    private val port: Int = 9100,
) : ByteTransport {

    private var socket: Socket? = null
    private var out: OutputStream? = null
    private var input: InputStream? = null

    override val isOpen: Boolean
        get() = socket?.isConnected == true && socket?.isClosed == false

    override fun open(timeoutMs: Long) {
        if (isOpen) return
        try {
            val s = Socket()
            s.tcpNoDelay = true
            s.connect(InetSocketAddress(host, port), timeoutMs.toInt())
            s.soTimeout = timeoutMs.toInt()
            socket = s
            out = s.getOutputStream()
            input = s.getInputStream()
        } catch (e: SocketTimeoutException) {
            throw PrinterException(ErrorCode.TIMEOUT, "Timeout connexion TCP $host:$port", e.message, retryable = true)
        } catch (e: Exception) {
            throw PrinterException(ErrorCode.CONNECTION_FAILED, "Connexion TCP échouée $host:$port", e.message, retryable = true)
        }
    }

    override fun write(bytes: ByteArray) {
        val o = out ?: throw PrinterException(ErrorCode.CONNECTION_FAILED, "Socket TCP non ouvert")
        try {
            // Envoi par chunks pour ménager les petits buffers d'imprimante.
            var offset = 0
            val chunk = 4096
            while (offset < bytes.size) {
                val len = minOf(chunk, bytes.size - offset)
                o.write(bytes, offset, len)
                o.flush()
                offset += len
            }
        } catch (e: Exception) {
            throw PrinterException(ErrorCode.PRINT_FAILED, "Écriture TCP échouée", e.message, retryable = true)
        }
    }

    override fun read(buffer: ByteArray, timeoutMs: Long): Int {
        val i = input ?: return -1
        return try {
            socket?.soTimeout = timeoutMs.toInt()
            i.read(buffer)
        } catch (e: SocketTimeoutException) {
            -1
        } catch (e: Exception) {
            -1
        }
    }

    override fun close() {
        try { out?.flush() } catch (_: Exception) {}
        try { input?.close() } catch (_: Exception) {}
        try { out?.close() } catch (_: Exception) {}
        try { socket?.close() } catch (_: Exception) {}
        socket = null; out = null; input = null
    }
}
